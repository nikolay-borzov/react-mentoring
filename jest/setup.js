import promiseFinally from 'promise.prototype.finally'

import Enzyme, { shallow, render, mount } from 'enzyme'
import Adapter from 'enzyme-adapter-react-16'

// Bring '.finally' to Promise (only for Node 8). Will be available in Node 10
promiseFinally.shim()

Enzyme.configure({ adapter: new Adapter() })

// Make Enzyme functions available in all test files without importing
global.shallow = shallow
global.render = render
global.mount = mount
