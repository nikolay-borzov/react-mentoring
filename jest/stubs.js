export const film = {
  id: 1,
  title: 'Film title',
  poster_path: 'http://via.placeholder.com/500x750',
  overview: 'Film overview',
  release_date: '2018-05-04',
  genres: ['some genre', 'another genre']
}

export const films = [film]
