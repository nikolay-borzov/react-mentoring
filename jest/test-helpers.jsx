import renderer from 'react-test-renderer'
import { shallow } from 'enzyme'
import { runSaga as runSagaOriginal } from 'redux-saga'
import configureMockStore from 'redux-mock-store'

/**
 * Creates test closure for snapshot testing
 * @deprecated Use itRendersCorrectlyShallow
 * @param {Function} getComponent Function returning component to test snapshot
 * @param {string} [testName='renders correctly']
 */
export const itRendersCorrectly = (
  getComponent,
  testName = 'renders correctly'
) => {
  it(testName, () => {
    const tree = renderer.create(getComponent()).toJSON()

    expect(tree).toMatchSnapshot()
  })
}

/**
 * Creates test closure for snapshot testing (using Enzyme shallow)
 * @param {Function} getComponent Function returning component to test snapshot
 * @param {string} [testName='renders correctly']
 */
export const itRendersCorrectlyShallow = (
  getComponent,
  testName = 'renders correctly'
) => {
  it(testName, () => {
    const wrapper = shallow(getComponent())

    expect(wrapper).toMatchSnapshot()
  })
}

/**
 * Creates test closure to check whether component contains another component
 * @param {Function} getComponent Function returning parent component
 * @param {string} componentDisplayName Child component display name
 * @param {Object} params Optional params
 * @param {number} [params.testName=`contains ${componentDisplayName} component`]
 * @param {number} [params.expectedCount=1] Expected number of children components
 * @param {Object} params.expectedProps First child component properties map
 */
export const itContainsComponent = (
  getComponent,
  componentDisplayName,
  { testName, expectedCount = 1, expectedProps } = {}
) => {
  it(testName || `contains ${componentDisplayName} component`, () => {
    const wrapper = shallow(getComponent())
    const children = wrapper.find(componentDisplayName)

    expect(children.length).toBe(expectedCount)
    if (expectedProps) {
      expect(children.first().props()).toEqual(expectedProps)
    }
  })
}

/**
 * Creates mocked store
 */
export const getMockStore = () => configureMockStore([])

/**
 * Sets relative URL
 * @param {string} url
 */
export const setUrl = url => {
  window.history.pushState({}, '', url)
}

/**
 * Runs saga and collects dispatched actions
 */
export const runSaga = (saga, state = {}, arg1 = undefined) => {
  const dispatched = []
  return runSagaOriginal(
    {
      dispatch: action => dispatched.push(action),
      getState: () => state
    },
    saga,
    arg1
  )
    .toPromise()
    .then(() => dispatched)
}
