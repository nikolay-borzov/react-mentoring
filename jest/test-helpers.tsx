import renderer from 'react-test-renderer'
import { shallow } from 'enzyme'
import { runSaga as runSagaOriginal } from 'redux-saga'
import configureMockStore from 'redux-mock-store'
import { AnyAction } from 'redux';

/**
 * Creates test closure for snapshot testing
 * @deprecated Use itRendersCorrectlyShallow
 */
export const itRendersCorrectly = (
  getComponent: Function,
  testName: string = 'renders correctly'
) => {
  it(testName, () => {
    const tree = renderer.create(getComponent()).toJSON()

    expect(tree).toMatchSnapshot()
  })
}

/**
 * Creates test closure for snapshot testing (using Enzyme shallow)
 * @param getComponent Function returning component to test snapshot
 */
export const itRendersCorrectlyShallow = (
  getComponent: () => React.ReactElement,
  testName: string = 'renders correctly'
) => {
  it(testName, () => {
    const wrapper = shallow(getComponent())

    expect(wrapper).toMatchSnapshot()
  })
}

interface ItContainsComponentParams {
  testName?: string
  expectedCount?: number
  expectedProps?: {}
}

/**
 * Creates test closure to check whether component contains another component
 * @param getComponent Function returning parent component
 * @param componentDisplayName Child component display name
 * @param params Optional params
 * @param [params.testName=`contains ${componentDisplayName} component`]
 * @param [params.expectedCount=1] Expected number of children components
 * @param params.expectedProps First child component properties map
 */
export const itContainsComponent = (
  getComponent: () => React.ReactElement,
  componentDisplayName: string,
  { testName = `contains ${componentDisplayName} component`, expectedCount = 1, expectedProps }: ItContainsComponentParams = {}
) => {
  it(testName, () => {
    const wrapper = shallow(getComponent())
    const children = wrapper.find(componentDisplayName)

    expect(children.length).toBe(expectedCount)
    if (expectedProps) {
      expect(children.first().props()).toEqual(expectedProps)
    }
  })
}

/**
 * Creates mocked store
 */
export const getMockStore = () => configureMockStore([])

/**
 * Sets relative URL
 * @param {string} url
 */
export const setUrl = (url: string) => {
  window.history.pushState({}, '', url)
}

/**
 * Runs saga and collects dispatched actions
 */
export const runSaga = (saga: any, state = {}, arg1: any = undefined) => {
  const dispatched: AnyAction[] = []
  return runSagaOriginal(
    {
      dispatch: (action: AnyAction) => dispatched.push(action),
      getState: () => state
    },
    saga,
    arg1
  )
    .toPromise()
    .then(() => dispatched)
}
