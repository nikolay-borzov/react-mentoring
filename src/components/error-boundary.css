.error-boundary details {
  whiteSpace: 'pre-wrap'
}