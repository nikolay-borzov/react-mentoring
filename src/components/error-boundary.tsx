import * as React from 'react'
import styled from 'styled-components'

interface ErrorBoundaryProps {
  children: React.ReactNode
}

interface ErrorInfo {
  componentStack: string
}

interface ErrorBoundaryState {
  hasError: boolean
  error?: Error
  errorInfo?: ErrorInfo
}

const ErrorBoundaryDetails = styled.details`
  whitespace: 'pre-wrap';
`

export class ErrorBoundary extends React.PureComponent<
  ErrorBoundaryProps,
  ErrorBoundaryState
> {
  state: ErrorBoundaryState = {
    hasError: false
  }

  componentDidCatch(error: Error, errorInfo: ErrorInfo) {
    this.setState({
      hasError: true,
      error,
      errorInfo
    })

    // TIP: Send error details to the server
  }

  render() {
    const { hasError, error, errorInfo } = this.state

    if (hasError) {
      return (
        <div className="error-boundary centered alt-background">
          <div className="padding-content">
            <h1>Something went wrong.</h1>

            <ErrorBoundaryDetails>
              <summary>
                Details
                <br />
              </summary>
              {error && error.toString()}
              <br />
              {errorInfo && errorInfo.componentStack}
            </ErrorBoundaryDetails>

            <p itemProp="telephone">
              Please try again later or contact support at&nbsp;
              <a href="tel:+46771793336" rel="nofollow">
                46 771 793 336
              </a>
            </p>
          </div>
        </div>
      )
    }

    return this.props.children
  }
}
