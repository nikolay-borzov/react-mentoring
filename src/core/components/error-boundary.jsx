import React from 'react'
import PropTypes from 'prop-types'

export class ErrorBoundary extends React.Component {
  static propTypes = {
    children: PropTypes.any.isRequired
  }

  constructor(props) {
    super(props)
    this.state = { hasError: false }
  }

  componentDidCatch(error, errorInfo) {
    this.setState({
      hasError: true,
      error: error,
      errorInfo: errorInfo
    })

    // TODO: Send error details to the server
  }

  render() {
    if (this.state.hasError) {
      return (
        <div className="centered alt-background">
          <div>
            <h1>Something went wrong.</h1>

            <details style={{ whiteSpace: 'pre-wrap' }}>
              <summary>
                Details<br />
              </summary>
              {this.state.error && this.state.error.toString()}
              <br />
              {this.state.errorInfo.componentStack}
            </details>

            <p itemProp="telephone">
              Please try again later or contact support at&nbsp;
              <a href="tel:+46771793336" rel="nofollow">
                46 771 793 336
              </a>
            </p>
          </div>
        </div>
      )
    }
    return this.props.children
  }
}
