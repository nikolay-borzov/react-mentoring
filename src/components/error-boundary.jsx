import React from 'react'
import PropTypes from 'prop-types'

import './error-boundary.css'

export class ErrorBoundary extends React.PureComponent {
  static propTypes = {
    children: PropTypes.any.isRequired
  }

  state = {
    hasError: false
  }

  componentDidCatch(error, errorInfo) {
    this.setState({
      hasError: true,
      error: error,
      errorInfo: errorInfo
    })

    // TIP: Send error details to the server
  }

  render() {
    if (this.state.hasError) {
      return (
        <div className="error-boundary centered alt-background">
          <div className="padding-content">
            <h1>Something went wrong.</h1>

            <details>
              <summary>
                Details<br />
              </summary>
              {this.state.error && this.state.error.toString()}
              <br />
              {this.state.errorInfo.componentStack}
            </details>

            <p itemProp="telephone">
              Please try again later or contact support at&nbsp;
              <a href="tel:+46771793336" rel="nofollow">
                46 771 793 336
              </a>
            </p>
          </div>
        </div>
      )
    }

    return this.props.children
  }
}
