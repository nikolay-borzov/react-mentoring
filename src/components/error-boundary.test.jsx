import React from 'react'
import { shallow } from 'enzyme'

import { itRendersCorrectly } from '../../jest/test-helpers'

import { ErrorBoundary } from './error-boundary'

describe('ErrorBoundary component', () => {
  // Mute error logs https://github.com/airbnb/enzyme/issues/1255
  jest.spyOn(window._virtualConsole, 'emit').mockImplementation(() => false)
  jest.spyOn(window.console, 'error').mockImplementation(() => false)

  itRendersCorrectly(
    () => <ErrorBoundary>Content</ErrorBoundary>,
    'renders children if no errors have happened'
  )

  itRendersCorrectly(() => {
    function ErroneousComponent() {
      throw new Error('Some genic error')
    }

    return (
      <ErrorBoundary>
        <ErroneousComponent />
      </ErrorBoundary>
    )
  }, 'renders correctly when an error has happened')

  it('should render children when passed in', () => {
    const wrapper = shallow(
      <ErrorBoundary>
        <div className="unique" />
      </ErrorBoundary>
    )

    expect(wrapper.contains(<div className="unique" />)).toBeTruthy()
  })
})
