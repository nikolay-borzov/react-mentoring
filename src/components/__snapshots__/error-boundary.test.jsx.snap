// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ErrorBoundary component renders children if no errors have happened 1`] = `"Content"`;

exports[`ErrorBoundary component renders correctly when an error has happened 1`] = `
<div
  className="error-boundary centered alt-background"
>
  <div
    className="padding-content"
  >
    <h1>
      Something went wrong.
    </h1>
    <details
      className="ErrorBoundaryDetails-ymaymh-0 jykFMN"
    >
      <summary>
        Details
        <br />
      </summary>
      Error: Some genic error
      <br />
      
    in ErroneousComponent
    in ErrorBoundary
    </details>
    <p
      itemProp="telephone"
    >
      Please try again later or contact support atÂ 
      <a
        href="tel:+46771793336"
        rel="nofollow"
      >
        46 771 793 336
      </a>
    </p>
  </div>
</div>
`;
